File: javascript/find_shortest_string.js
function findShortestString(arr) {
  // let smallest = arr[0]
  // for (const word of arr) {
  //   if (word.length < smallest.length) {
  //     smallest = word
  //   }
  // }
  // return smallest
  
  return arr.reduce((accum, currentValue) => {
    console.log(accum, currentValue);
    return currentValue.length < accum.length 
    ? currentValue 
    : accum;
   }, arr[0])
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestString(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestString(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestString(['flower', 'juniper', 'lily', 'dadelion']));

  // BENCHMARK HERE
}

module.exports = findShortestString;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: ruby/find_shortest_string.rb
require 'benchmark'

def find_shortest_string(arr)
  arr.min_by { |str| str.length }
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])

  # Don't forget to add your own!

  time1 = Benchmark.measure {
    find_shortest_string(['flower', 'juniper'])
  }
  time2 = Benchmark.measure {
    find_shortest_string([''])
  }
  time3 = Benchmark.measure {
    find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])
  }
  time4 = Benchmark.measure {
    find_shortest_string(%w(asdkfjh dkfjh askdfj ;asdlfjkh askl;dfjh asdfh jas;dfh jsdf asklfh jasklf asldf ;as jlaks jf;ash jdf;h jasdfkl;g jsd))
  }

  puts time1
  puts time2
  puts time3
  puts time4
end

# Please add your pseudocode to this file
# And a written explanation of your solution
